src/Tools/dotnet-stack/Symbolicate.cs

```diff
@@ -0,0 +1,419 @@
+        /// <returns></returns>
+        private static void Symbolicate(FileInfo[] inputPath, string searchDir, string output)
+        {
+            try
+            {
+                if (inputPath == null)
+                {
+                    throw new InvalidDataException("Required argument missing for command\n");
+                }
+
+                string input = string.Empty;
+                if (inputPath.Length > 1)
+                {
+                    for (int i = 0; i < inputPath.Length; i++)
+                    {
+                        Console.WriteLine($" {i + 1}. {inputPath[i]}");
+                    }
+                    Console.Write($"Select one of several stacktrace files [1-{inputPath.Length}]: ");
+                    input = inputPath[Int32.Parse(Console.ReadLine()) - 1].FullName;
+                }
+                else
+                {
+                    input = inputPath[0].FullName;
+                }
+                if (!File.Exists(input))
+                {
+                    throw new FileNotFoundException($"{input} file does not exist\n");
+                }
+
+                string[] search_paths = null;
+                if (searchDir != null)
+                {
+                    search_paths = searchDir.Split(";");
+                }
+                search_paths ??= new string[] { Directory.GetCurrentDirectory() };
+
+                Symbolicator(PdbToXmlConvert(search_paths), input, output);
+            }
+            catch (Exception e)
+            {
+                Console.Error.WriteLine(e.Message);
+            }
+        }
+
+        private static List<string> PdbToXmlConvert(string[] searchPaths)
+        {
+            List<string> peFiles = GrabFiles(searchPaths, "*.dll");
+            if (peFiles.Count == 0)
+            {
+                throw new FileNotFoundException("Assembly file not found\n");
+            }
+
+            List<string> pdbFiles = GrabFiles(searchPaths, "*.pdb");
+            if (pdbFiles.Count == 0)
+            {
+                throw new FileNotFoundException("PDB file not found\n");
+            }
+
+            Console.Out.Write("Converting pdb to xml...");
+            Console.SetCursorPosition(0, Console.CursorTop);
+
+            Directory.CreateDirectory(tempDirectory);
+
+            List<string> xmlList = new List<string>();
+            foreach (var pdbFile in pdbFiles)
+            {
+                foreach (var peFile in peFiles)
+                {
+                    if (Path.GetFileNameWithoutExtension(peFile) == Path.GetFileNameWithoutExtension(pdbFile))
+                    {
+                        string xmlPath = Path.Combine(tempDirectory, Path.GetFileName(Path.ChangeExtension(peFile, "xml")));
+                        GenXmlFromPdb(peFile, pdbFile, xmlPath);
+                        xmlList.Add(xmlPath);
+                        break;
+                    }
+                }
+            }
```

Comment:
I would expect we would only have one PDB per assembly. If you sorted both arrays before iterating and then used a `for(;;;)` loop to not lose your spot in the array, you would do less work.  As it is now, this could potentially iterate all the assembly files for every PDB.

Reply:
> I would expect we would only have one PDB per assembly.

Right. But https://github.com/dotnet/diagnostics/pull/2145#issuecomment-814633236
>In generally, CE products(mobile, wearable, TV, refrigerator, etc) do not have enough memory.
So, It is often deployed an application without a PDB file to reduce the deployment size or obfuscate.

I think PDB file per assembly may not exist. So, I generated XML file only when there is assembly and pdb of the same name. Do you have a better idea?

Reply:
I was only suggesting a way to match up the assembly list and the pdb list with less looping.

for example (in pseudo code):
```pseudo
function match_assembly_pdb:
	pairList <- []
	pdbList.sort()
	asmList.sort()
	j <- 0
	for i in pdbList.length: # assuming pdbList.length < asmList.length
		while j < asmList.length:
			if pdbList[i] == asmList[j]:
				pairList.add((pdbList[i], asmList[j]))
				break
			j++
```

Then the code only ever touches each element of each array once after the sort.

All that being said, I chatted with my team a bit and I think there is a simpler way of getting the symbols without converting to XML. Rather than use DiaSymReader, you can use the `System.Reflection.Metadata` APIs directly since those already know how to manipulate portable PDBs.

Here is an example from SOS on how you can use these APIs:
https://github.com/dotnet/diagnostics/blob/419db65453d353c654192b42fa5aa63785b02abc/src/SOS/SOS.Hosting/SymbolServiceWrapper.cs#L502-L547

Here is an example from the runtime on how to use those APIs (this one is a little simpler and handles caching parsed PDBs):
https://github.com/dotnet/runtime/blob/b1452c8fcab789dd9f2921c1479425b291111c5d/src/libraries/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.cs#L13-L94

Since non-portable PDBs are opt-in only and only work on Windows, I'm inclined to not add the extra headache of using DiaSym which ultimately wraps the `System.Reflection.Metadata` for portable PDBs anyway. This will help with caching parsed symbols, reduce dependencies, and decrease complexity by not requiring the conversion to XML phase.

Reply:
I want to emphasize the use of `System.Reflection.Metadata` here since that should simplify the logic and dependencies.

Reply:
> I want to emphasize the use of `System.Reflection.Metadata` here since that should simplify the logic and dependencies.

Ok. I'll check the ```System.Reflection.Metadata```